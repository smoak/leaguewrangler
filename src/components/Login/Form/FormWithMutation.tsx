import React, { FC } from 'react';
import { Omit } from '@material-ui/core';

import Form, { FormProps } from './Form';
import { Mutation, MutationOpts } from 'react-apollo';
import { CreateUserToken, CreateUserTokenVariables } from '../../../graphql/types/CreateUserToken';
import createUserToken from '../../../graphql/mutations/createUserToken';
import { setUserToken } from '../../../support/auth';

export type FormWithMutationProps = Omit<FormProps, 'onSignInClicked'>;

export class SignInMutation extends Mutation<CreateUserToken, CreateUserTokenVariables> {}

const onSignInCompleted: MutationOpts<CreateUserToken, CreateUserTokenVariables>['onCompleted'] = ({
  createUserToken: { token },
}) => {
  setUserToken(token);
  window.location.reload();
};

const FormWithMutation: FC<FormWithMutationProps> = props => {
  return (
    <SignInMutation mutation={createUserToken} onCompleted={onSignInCompleted}>
      {mutate => (
        <Form
          {...props}
          onSignInClicked={(username, password) => {
            mutate({
              variables: { username, password },
            });
          }}
        />
      )}
    </SignInMutation>
  );
};

export default FormWithMutation;
