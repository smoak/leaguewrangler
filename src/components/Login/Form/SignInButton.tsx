import { Button } from '@mui/material';
import { FC } from 'react';

interface SignInButtonProps {
  readonly isDisabled: boolean;
  onClick(): void;
}

const SignInButton: FC<SignInButtonProps> = ({ isDisabled, onClick }) => {
  return (
    <Button
      type="submit"
      fullWidth
      variant="contained"
      onClick={(e) => {
        e.preventDefault();
        onClick();
      }}
      sx={{ mt: 3, mb: 2 }}
      disabled={isDisabled}
    >
      Sign in
    </Button>
  );
};

export default SignInButton;
