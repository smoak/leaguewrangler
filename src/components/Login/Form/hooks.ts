import { useCallback, useState } from 'react';

export interface ControlledInput {
  readonly value: string;
  onChange: React.ChangeEventHandler<HTMLTextAreaElement | HTMLInputElement>;
}
type UseControlledInput = (defaultValue: string) => ControlledInput;
export const useControlledInput: UseControlledInput = defaultValue => {
  const [value, setValue] = useState(defaultValue);
  const onChange = useCallback<React.ChangeEventHandler<HTMLTextAreaElement | HTMLInputElement>>(e => {
    setValue(e.target.value);
  }, []);

  return {
    value,
    onChange,
  };
};
