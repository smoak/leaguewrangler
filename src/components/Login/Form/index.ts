export { default } from './FormWithMutation';
export * from './Form';
