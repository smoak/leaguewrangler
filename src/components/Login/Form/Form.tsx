import { useMutation } from '@apollo/client';
import { Checkbox, CircularProgress, FormControl, FormControlLabel, Input, InputLabel } from '@material-ui/core';
import { FC, useCallback } from 'react';
import { setUserToken } from 'support/auth';

import CREATE_USER_TOKEN from '../../../graphql/mutations/createUserToken';
import { CreateUserToken, CreateUserTokenVariables } from '../../../graphql/types/CreateUserToken';

import SignInButton from './SignInButton';
import { useControlledInput } from './hooks';

export interface FormProps {
  readonly formClassname: string;
  readonly submitButtonClassname: string;
}

const Form: FC<FormProps> = ({ submitButtonClassname, formClassname }) => {
  const { value: username, onChange: onUsernameChange } = useControlledInput('');
  const { value: password, onChange: onPasswordChange } = useControlledInput('');
  const [createUserToken, { data, error, loading }] = useMutation<CreateUserToken, CreateUserTokenVariables>(
    CREATE_USER_TOKEN
  );

  const onSignInClicked = useCallback(() => {
    createUserToken({ variables: { username, password } });
  }, [createUserToken, username, password]);

  if (data && data.createUserToken) {
    setUserToken(data.createUserToken.token);
    window.location.reload();
  }

  const button = loading ? (
    <CircularProgress color="secondary" />
  ) : (
    <SignInButton onClick={onSignInClicked} isDisabled={!username || !password} className={submitButtonClassname} />
  );

  return (
    <div className={formClassname}>
      <FormControl margin="normal" required fullWidth error={!!error}>
        <InputLabel htmlFor="username">Username</InputLabel>
        <Input
          id="username"
          name="username"
          autoComplete="username"
          autoFocus={true}
          value={username}
          onChange={onUsernameChange}
        />
      </FormControl>
      <FormControl margin="normal" required fullWidth error={!!error}>
        <InputLabel htmlFor="password">Password</InputLabel>
        <Input
          name="password"
          type="password"
          id="password"
          autoComplete="current-password"
          value={password}
          onChange={onPasswordChange}
          data-testid="password-textbox"
        />
      </FormControl>
      <FormControlLabel control={<Checkbox value="remember" color="primary" />} label="Remember me" />
      {button}
    </div>
  );
};

export default Form;
