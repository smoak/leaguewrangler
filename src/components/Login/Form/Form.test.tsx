import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import { InputProps } from '@material-ui/core/Input';
import { Button } from '@material-ui/core';

import Form from './Form';

describe('Form', () => {
  let component: ShallowWrapper;
  let onSignInClicked: jest.Mock;

  beforeEach(() => {
    onSignInClicked = jest.fn().mockName('onSignInClicked');
    component = shallow(<Form formClassname="form" submitButtonClassname="submit" onSignInClicked={onSignInClicked} />);
  });

  it('renders', () => {
    expect(component).toMatchSnapshot();
  });

  describe('when username is changed', () => {
    beforeEach(() => {
      const event = { target: { value: 'foo' } };
      component = shallow(
        <Form formClassname="form" submitButtonClassname="submit" onSignInClicked={onSignInClicked} />
      );
      component.find('#username').simulate('change', event);
    });

    it('updates the input value', () => {
      expect(component.find('#username').prop<InputProps>('value')).toEqual('foo');
    });
  });

  describe('when password is changed', () => {
    beforeEach(() => {
      const event = { target: { value: 'bar' } };
      component = shallow(
        <Form formClassname="form" submitButtonClassname="submit" onSignInClicked={onSignInClicked} />
      );
      component.find('#password').simulate('change', event);
    });

    it('updates the input value', () => {
      expect(component.find('#password').prop<InputProps>('value')).toEqual('bar');
    });
  });

  describe('when the user clicks the sign in button', () => {
    beforeEach(() => {
      component = shallow(
        <Form formClassname="form" submitButtonClassname="submit" onSignInClicked={onSignInClicked} />
      );
      const button = component.find(Button);
      const event = { preventDefault: jest.fn() };
      button.simulate('click', event);
    });

    it('calls onSignInClicked with the username and password', () => {
      expect(onSignInClicked).toHaveBeenCalledWith('', '');
    });
  });
});
