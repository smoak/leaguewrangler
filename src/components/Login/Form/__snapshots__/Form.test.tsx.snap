// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders 1`] = `
<div
  className="form"
>
  <WithStyles(FormControl)
    fullWidth={true}
    margin="normal"
    required={true}
  >
    <WithStyles(WithFormControlContext(InputLabel))
      htmlFor="username"
    >
      Username
    </WithStyles(WithFormControlContext(InputLabel))>
    <WithStyles(Input)
      autoComplete="username"
      autoFocus={true}
      id="username"
      name="username"
      onChange={[Function]}
      value=""
    />
  </WithStyles(FormControl)>
  <WithStyles(FormControl)
    fullWidth={true}
    margin="normal"
    required={true}
  >
    <WithStyles(WithFormControlContext(InputLabel))
      htmlFor="password"
    >
      Password
    </WithStyles(WithFormControlContext(InputLabel))>
    <WithStyles(Input)
      autoComplete="current-password"
      id="password"
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </WithStyles(FormControl)>
  <WithStyles(WithFormControlContext(FormControlLabel))
    control={
      <WithStyles(Checkbox)
        color="primary"
        value="remember"
      />
    }
    label="Remember me"
  />
  <WithStyles(Button)
    className="submit"
    color="primary"
    fullWidth={true}
    onClick={[Function]}
    type="submit"
    variant="contained"
  >
    Sign in
  </WithStyles(Button)>
</div>
`;
